File: csharp/HPKVClient/HPKVClient.cs
using System;
using System.Collections.Generic;
using System.Net;
using System.Text.Json;
using System.Threading.Tasks;
using RestSharp;
using HPKVClient.Models;

namespace HPKVClient
{
    /// <summary>
    /// Client for interacting with the HPKV REST API
    /// </summary>
    public class HPKVClient
    {
        private readonly RestClient _client;
        private readonly string _apiKey;
        private readonly JsonSerializerOptions _jsonOptions;

        /// <summary>
        /// Creates a new HPKV client
        /// </summary>
        /// <param name="baseUrl">The base URL for the HPKV API</param>
        /// <param name="apiKey">Your HPKV API key</param>
        public HPKVClient(string baseUrl, string apiKey)
        {
            if (string.IsNullOrEmpty(baseUrl))
                throw new ArgumentException("baseUrl is required", nameof(baseUrl));
            if (string.IsNullOrEmpty(apiKey))
                throw new ArgumentException("apiKey is required", nameof(apiKey));

            _apiKey = apiKey;
            
            var options = new RestClientOptions(baseUrl)
            {
                ThrowOnAnyError = false
            };
            
            _client = new RestClient(options);
            _jsonOptions = new JsonSerializerOptions
            {
                PropertyNameCaseInsensitive = true
            };
        }

        /// <summary>
        /// Insert or update a record
        /// </summary>
        /// <param name="key">Key to store</param>
        /// <param name="value">Value to store</param>
        /// <param name="partialUpdate">Whether to perform a partial update</param>
        /// <returns>The API response</returns>
        public async Task<OperationResponse> SetAsync(string key, object value, bool partialUpdate = false)
        {
            var request = new RestRequest("/record", Method.Post);
            request.AddHeader("Content-Type", "application/json");
            request.AddHeader("x-api-key", _apiKey);

            var payload = new SetRecordRequest
            {
                Key = Uri.EscapeDataString(key),
                Value = value is string ? value : JsonSerializer.Serialize(value),
                PartialUpdate = partialUpdate
            };

            request.AddJsonBody(payload);

            var response = await _client.ExecuteAsync(request);
            HandleErrors(response);

            return JsonSerializer.Deserialize<OperationResponse>(response.Content, _jsonOptions);
        }

        /// <summary>
        /// Get a record by key
        /// </summary>
        /// <param name="key">Key to retrieve</param>
        /// <returns>The record</returns>
        public async Task<GetRecordResponse> GetAsync(string key)
        {
            var request = new RestRequest($"/record/{Uri.EscapeDataString(key)}", Method.Get);
            request.AddHeader("x-api-key", _apiKey);

            var response = await _client.ExecuteAsync(request);
            HandleErrors(response);

            return JsonSerializer.Deserialize<GetRecordResponse>(response.Content, _jsonOptions);
        }

        /// <summary>
        /// Delete a record
        /// </summary>
        /// <param name="key">Key to delete</param>
        /// <returns>The API response</returns>
        public async Task<OperationResponse> DeleteAsync(string key)
        {
            var request = new RestRequest($"/record/{Uri.EscapeDataString(key)}", Method.Delete);
            request.AddHeader("x-api-key", _apiKey);

            var response = await _client.ExecuteAsync(request);
            HandleErrors(response);

            return JsonSerializer.Deserialize<OperationResponse>(response.Content, _jsonOptions);
        }

        /// <summary>
        /// Increment or decrement a numeric value
        /// </summary>
        /// <param name="key">Key to increment/decrement</param>
        /// <param name="increment">Value to add (positive) or subtract (negative)</param>
        /// <returns>The API response with the new value</returns>
        public async Task<IncrementResponse> IncrementAsync(string key, int increment = 1)
        {
            var request = new RestRequest("/record/atomic", Method.Post);
            request.AddHeader("Content-Type", "application/json");
            request.AddHeader("x-api-key", _apiKey);

            var payload = new IncrementRequest
            {
                Key = Uri.EscapeDataString(key),
                Increment = increment
            };

            request.AddJsonBody(payload);

            var response = await _client.ExecuteAsync(request);
            HandleErrors(response);

            return JsonSerializer.Deserialize<IncrementResponse>(response.Content, _jsonOptions);
        }

        /// <summary>
        /// Query records within a key range
        /// </summary>
        /// <param name="startKey">Starting key (inclusive)</param>
        /// <param name="endKey">Ending key (inclusive)</param>
        /// <param name="limit">Maximum number of records to return</param>
        /// <returns>Records in the range</returns>
        public async Task<RangeQueryResponse> QueryAsync(string startKey, string endKey, int limit = 100)
        {
            var request = new RestRequest("/records", Method.Get);
            request.AddHeader("x-api-key", _apiKey);
            request.AddParameter("startKey", Uri.EscapeDataString(startKey));
            request.AddParameter("endKey", Uri.EscapeDataString(endKey));
            request.AddParameter("limit", limit);

            var response = await _client.ExecuteAsync(request);
            HandleErrors(response);

            return JsonSerializer.Deserialize<RangeQueryResponse>(response.Content, _jsonOptions);
        }

        /// <summary>
        /// Handle API errors
        /// </summary>
        /// <param name="response">The response to check for errors</param>
        private void HandleErrors(RestResponse response)
        {
            if (!response.IsSuccessful)
            {
                BaseResponse errorResponse = null;
                string message;

                try
                {
                    if (!string.IsNullOrEmpty(response.Content))
                    {
                        errorResponse = JsonSerializer.Deserialize<BaseResponse>(response.Content, _jsonOptions);
                    }
                }
                catch { /* Ignore JSON parse error */ }

                if (errorResponse != null && !string.IsNullOrEmpty(errorResponse.Error))
                {
                    message = errorResponse.Error;
                }
                else
                {
                    message = $"HTTP error {(int)response.StatusCode}";
                }

                var exception = new HPKVException(message)
                {
                    StatusCode = response.StatusCode,
                    ResponseData = errorResponse
                };

                throw exception;
            }
        }
    }

    /// <summary>
    /// Exception thrown for HPKV API errors
    /// </summary>
    public class HPKVException : Exception
    {
        /// <summary>
        /// HTTP status code of the response
        /// </summary>
        public HttpStatusCode StatusCode { get; set; }

        /// <summary>
        /// Data returned in the error response
        /// </summary>
        public BaseResponse ResponseData { get; set; }

        /// <summary>
        /// Creates a new HPKVException
        /// </summary>
        /// <param name="message">Error message</param>
        public HPKVException(string message) : base(message)
        {
        }
    }
} 

File: java/src/main/java/io/hpkv/client/HPKVClient.java
package io.hpkv.client;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.TextNode;
import io.hpkv.client.models.*;
import okhttp3.*;

import java.io.IOException;
import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.TimeUnit;

/**
 * Client for interacting with the HPKV REST API
 */
public class HPKVClient {
    private final OkHttpClient httpClient;
    private final String baseUrl;
    private final String apiKey;
    private final ObjectMapper objectMapper;
    private final MediaType JSON = MediaType.parse("application/json; charset=utf-8");

    /**
     * Creates a new HPKV client
     *
     * @param baseUrl The base URL for the HPKV API
     * @param apiKey  Your HPKV API key
     */
    public HPKVClient(String baseUrl, String apiKey) {
        if (baseUrl == null || baseUrl.trim().isEmpty()) {
            throw new IllegalArgumentException("baseUrl is required");
        }
        if (apiKey == null || apiKey.trim().isEmpty()) {
            throw new IllegalArgumentException("apiKey is required");
        }

        this.baseUrl = baseUrl.endsWith("/") ? baseUrl.substring(0, baseUrl.length() - 1) : baseUrl;
        this.apiKey = apiKey;

        this.httpClient = new OkHttpClient.Builder()
                .connectTimeout(10, TimeUnit.SECONDS)
                .readTimeout(30, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .build();

        this.objectMapper = new ObjectMapper();
    }

    /**
     * Insert or update a record
     *
     * @param key          Key to store
     * @param value        Value to store
     * @param partialUpdate Whether to perform a partial update
     * @return The API response
     * @throws HPKVException If an API error occurs
     * @throws IOException   If a communication error occurs
     */
    public OperationResponse set(String key, Object value, boolean partialUpdate) throws HPKVException, IOException {
        // If value is a string, use it directly, otherwise serialize it to JSON
        Object valueToStore;
        if (value instanceof String) {
            valueToStore = value;
        } else {
            valueToStore = objectMapper.writeValueAsString(value);
        }

        SetRecordRequest request = new SetRecordRequest();
        request.setKey(encodeKey(key));
        request.setValue(valueToStore);
        request.setPartialUpdate(partialUpdate);

        String jsonBody = objectMapper.writeValueAsString(request);
        RequestBody body = RequestBody.create(jsonBody, JSON);

        Request httpRequest = new Request.Builder()
                .url(baseUrl + "/record")
                .addHeader("x-api-key", apiKey)
                .post(body)
                .build();

        try (Response response = httpClient.newCall(httpRequest).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";
            
            if (!response.isSuccessful()) {
                throw handleErrorResponse(response.code(), responseBody);
            }

            return objectMapper.readValue(responseBody, OperationResponse.class);
        }
    }

    /**
     * Get a record by key
     *
     * @param key Key to retrieve
     * @return The record
     * @throws HPKVException If an API error occurs
     * @throws IOException   If a communication error occurs
     */
    public GetRecordResponse get(String key) throws HPKVException, IOException {
        Request request = new Request.Builder()
                .url(baseUrl + "/record/" + encodeKey(key))
                .addHeader("x-api-key", apiKey)
                .get()
                .build();

        try (Response response = httpClient.newCall(request).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";

            if (!response.isSuccessful()) {
                throw handleErrorResponse(response.code(), responseBody);
            }

            return objectMapper.readValue(responseBody, GetRecordResponse.class);
        }
    }

    /**
     * Delete a record
     *
     * @param key Key to delete
     * @return The API response
     * @throws HPKVException If an API error occurs
     * @throws IOException   If a communication error occurs
     */
    public OperationResponse delete(String key) throws HPKVException, IOException {
        Request request = new Request.Builder()
                .url(baseUrl + "/record/" + encodeKey(key))
                .addHeader("x-api-key", apiKey)
                .delete()
                .build();

        try (Response response = httpClient.newCall(request).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";

            if (!response.isSuccessful()) {
                throw handleErrorResponse(response.code(), responseBody);
            }

            return objectMapper.readValue(responseBody, OperationResponse.class);
        }
    }

    /**
     * Increment or decrement a numeric value
     *
     * @param key       Key to increment/decrement
     * @param increment Value to add (positive) or subtract (negative)
     * @return The API response with the new value
     * @throws HPKVException If an API error occurs
     * @throws IOException   If a communication error occurs
     */
    public IncrementResponse increment(String key, int increment) throws HPKVException, IOException {
        IncrementRequest request = new IncrementRequest();
        request.setKey(encodeKey(key));
        request.setIncrement(increment);

        String jsonBody = objectMapper.writeValueAsString(request);
        RequestBody body = RequestBody.create(jsonBody, JSON);

        Request httpRequest = new Request.Builder()
                .url(baseUrl + "/record/atomic")
                .addHeader("x-api-key", apiKey)
                .post(body)
                .build();

        try (Response response = httpClient.newCall(httpRequest).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";

            if (!response.isSuccessful()) {
                throw handleErrorResponse(response.code(), responseBody);
            }

            return objectMapper.readValue(responseBody, IncrementResponse.class);
        }
    }

    /**
     * Query records within a key range
     *
     * @param startKey Starting key (inclusive)
     * @param endKey   Ending key (inclusive)
     * @param limit    Maximum number of records to return
     * @return Records in the range
     * @throws HPKVException If an API error occurs
     * @throws IOException   If a communication error occurs
     */
    public RangeQueryResponse query(String startKey, String endKey, int limit) throws HPKVException, IOException {
        HttpUrl.Builder urlBuilder = HttpUrl.parse(baseUrl + "/records").newBuilder()
                .addQueryParameter("startKey", URLEncoder.encode(startKey, StandardCharsets.UTF_8))
                .addQueryParameter("endKey", URLEncoder.encode(endKey, StandardCharsets.UTF_8))
                .addQueryParameter("limit", String.valueOf(limit));

        Request request = new Request.Builder()
                .url(urlBuilder.build())
                .addHeader("x-api-key", apiKey)
                .get()
                .build();

        try (Response response = httpClient.newCall(request).execute()) {
            String responseBody = response.body() != null ? response.body().string() : "";

            if (!response.isSuccessful()) {
                throw handleErrorResponse(response.code(), responseBody);
            }

            return objectMapper.readValue(responseBody, RangeQueryResponse.class);
        }
    }

    /**
     * Encode a key for URL path inclusion
     *
     * @param key Key to encode
     * @return Encoded key
     */
    private String encodeKey(String key) {
        return URLEncoder.encode(key, StandardCharsets.UTF_8);
    }

    /**
     * Handle API errors
     *
     * @param statusCode HTTP status code
     * @param responseBody Response body as string
     * @return An exception with error details
     */
    private HPKVException handleErrorResponse(int statusCode, String responseBody) {
        String message;
        BaseResponse errorResponse = null;

        try {
            if (responseBody != null && !responseBody.isEmpty()) {
                errorResponse = objectMapper.readValue(responseBody, BaseResponse.class);
                if (errorResponse.getError() != null && !errorResponse.getError().isEmpty()) {
                    message = errorResponse.getError();
                } else {
                    message = "HTTP error " + statusCode;
                }
            } else {
                message = "HTTP error " + statusCode;
            }
        } catch (JsonProcessingException e) {
            message = "HTTP error " + statusCode + " (Invalid response format)";
        }

        return new HPKVException(message, statusCode, errorResponse);
    }
} 